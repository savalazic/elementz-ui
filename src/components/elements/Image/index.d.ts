export { ImageProps, BackgroundImageProps } from './Image';
