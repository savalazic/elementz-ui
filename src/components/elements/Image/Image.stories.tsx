import React from 'react';

import { storiesOf } from '@storybook/react';

import { infoConfig } from '../../../utils/info-config';

import { Image, BackgroundImage } from './Image';
import { Text } from '../Text';

const stories = storiesOf('Elements|Image', module);

stories.addParameters(infoConfig);

stories
  .add('default', () => {
    return (
      <Image
        src="https://source.unsplash.com/600x400/daily/?nature"
        alt="Image from unsplash"
      />
    );
  })
  .add('rounded', () => {
    return (
      <Image
        src="https://source.unsplash.com/600x400/daily/?nature"
        alt="Image from unsplash"
        rounded="30px"
      />
    );
  })
  .add('circle', () => {
    return (
      <Image
        src="https://source.unsplash.com/600x400/daily/?nature"
        alt="Image from unsplash"
        circle="200px"
      />
    );
  })
  .add('square', () => {
    return (
      <Image
        src="https://source.unsplash.com/600x400/daily/?people"
        alt="Image from unsplash"
        square="200px"
      />
    );
  })
  .add('as background image', () => {
    return (
      <BackgroundImage
        src="https://source.unsplash.com/600x400/daily/?nature"
        style={{ padding: 40, color: 'white' }}
      >
        <Text tag="h2">Heading</Text>
        <Text>
          Lorem ipsum dolor sit amet consectetur, adipisicing elit. Facere
          veritatis non natus ipsum soluta voluptas, tempore voluptate saepe
          harum vel ea, recusandae eveniet? Beatae ipsa unde cupiditate iusto
          accusamus aspernatur!
        </Text>
      </BackgroundImage>
    );
  });
