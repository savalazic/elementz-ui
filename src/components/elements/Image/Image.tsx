import React from 'react';
import styled from 'styled-components';

export interface ImageProps extends React.ImgHTMLAttributes<HTMLImageElement> {
  /** image source */
  src: string;
  /** image alternate text */
  alt: string;
  /** rounded */
  rounded?: string;
  /** circle */
  circle?: string;
  /** square */
  square?: string;
}

export interface BackgroundImageProps
  extends React.HTMLAttributes<HTMLDivElement> {
  /** image source */
  src: string;
  /** @default center */
  /** image position */
  backgroundPosition?: 'center' | 'top' | 'right' | 'bottom' | 'left';
  /** children */
  children: React.ReactNode;
}

const StyledImage = styled.img<ImageProps>`
  max-width: 100%;
  object-fit: cover;

  border-radius: ${props => props.circle && '50%'};
  border-radius: ${props => props.rounded};
  width: ${props => props.circle || props.square};
  height: ${props => props.circle || props.square};
`;

const StyledBackgroundImage = styled.div<BackgroundImageProps>`
  background-image: url(${props => props.src});
  background-size: cover;
  background-repeat: no-repeat;
`;

export const Image = ({
  src,
  alt,
  rounded,
  circle,
  square,
  ...rest
}: ImageProps) => (
  <StyledImage
    src={src}
    alt={alt}
    circle={circle}
    rounded={rounded}
    square={square}
    {...rest}
  />
);

export const BackgroundImage = ({
  src,
  backgroundPosition = 'center',
  children,
  ...rest
}: BackgroundImageProps) => (
  <StyledBackgroundImage src={src} {...rest}>
    {children}
  </StyledBackgroundImage>
);
