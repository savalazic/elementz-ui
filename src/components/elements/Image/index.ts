export { Image, BackgroundImage } from './Image';
