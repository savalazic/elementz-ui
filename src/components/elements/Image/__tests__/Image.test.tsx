import 'jest-dom/extend-expect';
import React from 'react';
import { render } from 'react-testing-library';

import { Image, BackgroundImage } from '../Image';

describe('Image component', () => {
  const mockedSrc = 'https://source.unsplash.com/600x400/daily/?nature';
  const mockedAlt = 'mocked alt';

  it('should match default snapshot', () => {
    const { container } = render(<Image src={mockedSrc} alt={mockedAlt} />);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with rounded prop', () => {
    const { container } = render(
      <Image src={mockedSrc} alt={mockedAlt} rounded="4px" />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with circle prop', () => {
    const { container } = render(
      <Image src={mockedSrc} alt={mockedAlt} circle="50px" />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with square prop', () => {
    const { container } = render(
      <Image src={mockedSrc} alt={mockedAlt} square="40px" />,
    );
    expect(container).toMatchSnapshot();
  });
});

describe('BackgroundImage component', () => {
  const mockedSrc = 'https://source.unsplash.com/600x400/daily/?nature';
  const componentChildren = (
    <div>
      <p>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo, sequi.
      </p>
    </div>
  );

  it('should match default snapshot', () => {
    const { container } = render(
      <BackgroundImage src={mockedSrc}>{componentChildren}</BackgroundImage>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with backgroundPosition prop of `top`', () => {
    const { container } = render(
      <BackgroundImage src={mockedSrc} backgroundPosition="top">
        {componentChildren}
      </BackgroundImage>,
    );
    expect(container).toMatchSnapshot();
  });
});
