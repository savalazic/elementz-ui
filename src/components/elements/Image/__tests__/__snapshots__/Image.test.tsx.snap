// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackgroundImage component should match default snapshot 1`] = `
<div>
  <div
    class="sc-bwzfXH jNCWaz"
    src="https://source.unsplash.com/600x400/daily/?nature"
  >
    <div>
      <p>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo, sequi.
      </p>
    </div>
  </div>
</div>
`;

exports[`BackgroundImage component should match snapshot with backgroundPosition prop of \`top\` 1`] = `
<div>
  <div
    class="sc-bwzfXH jNCWaz"
    src="https://source.unsplash.com/600x400/daily/?nature"
  >
    <div>
      <p>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Quo, sequi.
      </p>
    </div>
  </div>
</div>
`;

exports[`Image component should match default snapshot 1`] = `
<div>
  <img
    alt="mocked alt"
    class="sc-bdVaJa bZxZdl"
    src="https://source.unsplash.com/600x400/daily/?nature"
  />
</div>
`;

exports[`Image component should match snapshot with circle prop 1`] = `
<div>
  <img
    alt="mocked alt"
    class="sc-bdVaJa RzhRA"
    src="https://source.unsplash.com/600x400/daily/?nature"
  />
</div>
`;

exports[`Image component should match snapshot with rounded prop 1`] = `
<div>
  <img
    alt="mocked alt"
    class="sc-bdVaJa jLXoOx"
    src="https://source.unsplash.com/600x400/daily/?nature"
  />
</div>
`;

exports[`Image component should match snapshot with square prop 1`] = `
<div>
  <img
    alt="mocked alt"
    class="sc-bdVaJa kwsylX"
    src="https://source.unsplash.com/600x400/daily/?nature"
  />
</div>
`;
